### PR TITLE
Update dfm

### PR DESCRIPTION
**Pull request by koheiw**
Call `dfm.tokens_xptr()` from `dfm.tokens()` for #2249. There are many changes and tests accordingly.

### CODE REVIEW

**Comment by koheiw** [comment]
@kbenoit can you review this?

File: R/dfm.R
Comment (kbenoit) [inline R/dfm.R:56]: These should be marked `.Defunct` rather than just removed. I am considering using the **lifecycle** package for deprecations and removals, see https://r-pkgs.org/lifecycle.html#sec-lifecycle-breaking-change-pros-cons, as this works really well for informing users of exactly what was removed and when. This will save a lot of clueless SO questions because it's more informative.

Comment (kbenoit) [reply R/dfm.R:56]: I agree we want to remove them in v4, the issue is simply how. No real need to do this here. Although I know how tempting it is. I promise a major cleanup is coming!

**Review by kbenoit — COMMENTED** [review]
This is a nice improvement and makes the underlying engine all based on xptr. It also removes the option of not doing that. Since the dfm is the same object, however, this should not affect compatibility, since a v3 dfm and a v4 dfm will be the same objects. (But do correct me if I'm wrong here!)

Two comments.

1. As I always emphasise, deprecations should be separate from functionality changes, whenever they are not integral to functionality change. Pruning is satisfying (I know!) but creates untold headaches for me in particular. The most time-consuming part of releasing v3 was me notifying other package maintainers of breaking changes and often, issuing my own PRs to fix code in their packages that would be broken by even our warnings. There is also the issue of the user base and their code. I agree we will defunct anything deprecated in v3, but there are some nifty tools for doing this now, and they are far more informative than just `.Defunct()`. (See https://r-pkgs.org/lifecycle.html) So I prefer to do these in a separate series of PRs because this is the most delicate surgery. That means I'd prefer the removals not be removed here. There is a lot that is good that you tidied here, including some tests we suppressed the warnings on that I agree should be removed here, but not simply removing code for v3 deprecations.

2. I still don't like the nomenclature of tokens_xptr. It is not consistent with our single-word object names, and the use of the underscore. It's not a function we have to call, I know, so I'm not 100% convinced it's a problem, but it's inconsistent - and consistency is one of the beautiful things about **quanteda**.

So I don't like seeing this for instance:
```
Error: dfm() only works on dfm, tokens, tokens_xptr objects.
```

Can we name this child class of tokens something that is a single word? `tokensx` or even just remove the underscore so it's `tokensxptr`? (I prefer the first.)

But it's not a deal-breaker.

**Comment by koheiw** [comment]
Thanks for reviewing. I decided to remove deprecated functions to to reduce the workload. It does not make sense to write code for functions that will be removed. Introduction of the `tokens_xptr` object as you can see in the number of commits that I made in the last three months.

**Comment by koheiw** [comment]
There should not be any function that tokenize texts on the fly like `bootstrap_dfm()` and `nsentecens()` etc. This is why I removed. Uses should pass a DFM to `bootstrap_dfm()`.

**Comment by koheiw** [comment]
More generally, we should reduced the number of functions more aggressively. It is becoming too hard to maintain. I am the only person who is doing the substantive development. Otherwise, I will really create **quanteda.core** package and leave functions that I think redundant with you to maintain.

File: R/data-documentation.R
Comment (koheiw) [reply R/data-documentation.R:128]: Please do if you want to.

File: R/bootstrap_dfm.R
Comment (koheiw) [reply R/bootstrap_dfm.R:35]: We discussed this two years ago. Users should pass DFM where documents are sentences. I want to removed `bootstrap_dfm()` entirely because I don't see any use case for me and it causes triggers this type of arguments often.

File: tests/testthat/test-dfm_group.R
Comment (koheiw) [reply tests/testthat/test-dfm_group.R:37]: `dfm()` no longer has the `group` arguments.

File: tests/testthat/test-dfm_lookup.R
Comment (koheiw) [reply tests/testthat/test-dfm_lookup.R:16]: Please do if yo want to.

**Comment by koheiw** [comment]
Please also note that despite the removed tests, the test coverage is higher significantly. We should keep removing redundant tests and examples too. I don't delete important tests.

**Comment by kbenoit** [comment]
@koheiw I'm so deeply impressed by how much time that you are putting into the package development - and the spectacular performance gains they have brought - that I almost messaged you out of concern whether you still have a job! (I hope so...)

If there are functions you think should be removed, propose them in an issue. But please don't remove them from a performance-related PR because this slows down the assessment of the PR. We can discuss and remove them separately. Anything that generates errors or warnings in rev deps is one of the hardest things to test for CRAN compliance. And most time-consuming to fix. I don't mind doing it but I prefer to minimise the post-surgery recovery time!

**Comment by koheiw** [comment]
I choose the name `toknes_xptr` for the following reasons when I started developing.

1. The new tokens object inherits properties of `tokens` and `externalptr` objects, so the class label could be `c("externalptr", "tokens")`, but it triggers methods for `externalptr` when I don't want to. If `c("tokens", "externalptr")`, I have to check the second class label in all the S3 methods for `tokens`.
2. I need a special class label for the new label that tell the users the nature of the object like "tokens_externalptr". It could be "tokensx", "xtokens" to make it short, but then I have to make a functions called `tokensx_*()` or `xtokens_*()`. It could be `tokens-xptr` or `tokens.xptr` but we know that the underscore is the best for S3 class labels.
3. It cannot be "tokens2" because "tokens" will co-exist with the new object.
4. We already have class labels with underscores like "textstat_proxy", "textstat_simil_sparse".

**Comment by koheiw** [comment]
By the way, this is not a PR for performance improvement. I fixed the C++ function on this branch because replacing R code with C++ code broke many tests.

**Comment by koheiw** [comment]
I decided to develop v4.0 because the data that I analyze on my job is getting larger and larger.

**Comment by kbenoit** [comment]
OK, all good points, let's do this then. 
1. We stick with the class name, since what you say makes sense, and I was already thinking about the underscored classes in textstats and textmodels as precedents, so let's stick with `tokens_xptr`.
2. I'll restore the removed items that are not integral to the PR, but remove them in a separate PR so that I can test their effects, and not hold up things further. Then we can merge this PR and move on to the rest of v4 plan.

File: R/dfm.R
Comment (kbenoit) [reply R/dfm.R:56]: See [7fdea44](https://github.com/quanteda/quanteda/pull/2251/commits/7fdea44cf9de1c24afb00e57c9516f26c8d6b909), this Defuncts them in the most minimal way possible. Also adds an argument to `check_class()` so that defunct methods are not listed as valid when the warning is printed for invalid object class inputs.

**Review by kbenoit — APPROVED** [review]
I'm going to start running `revdep_check()` with each PR but let's merge it first and I'll start the process then.

File: R/dfm.R
Comment (kbenoit) [reply R/dfm.R:56]: @koheiw if you're happy with this for now, you or I can mark this conversation as resolved and then merge the PR.

**Pull request merged**
